File: library_py/data_structure/unionfind.py
class UnionFind:
    def __init__(self, n):
        self.n = n
        self.parents = [-1] * n

    def unite(self, x, y):
        x, y = self.find(x), self.find(y)
        if x != y:
            if self.parents[x] > self.parents[y]:
                x, y = y, x
            self.parents[x] += self.parents[y]
            self.parents[y] = x
            return True

        return False

    def find(self, x):
        y = x
        while self.parents[x] >= 0:
            x = self.parents[x]
        while y != x:
            t = y
            y = self.parents[y]
            self.parents[t] = x
        return x

    def size(self, x):
        return -self.parents[self.find(x)]

    def same(self, x, y):
        return self.find(x) == self.find(y)

File: library_cpp/data_structure/unionfind.cpp
#include<bits/stdc++.h>


class UnionFind {
public:
    int n;
    std::vector<int> parents;

    UnionFind(int n) {
        this->n = n;
        parents = std::vector<int>(n, -1);
    }

    bool unite(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) {
            if (parents[x] > parents[y]) {
                std::swap(x, y);
            }
            parents[x] += parents[y];
            parents[y] = x;
            return true;
        }
        return false;
    }

    int find(int x) {
        int y = x;
        while (parents[x] >= 0) {
            x = parents[x];
        }
        while (y != x) {
            int t = y;
            y = parents[y];
            parents[t] = x;
        }
        return x;
    }

    int size(int x) {
        return -parents[find(x)];
    }

    bool same(int x, int y) {
        return find(x) == find(y);
    }
};
